File: 26-Visitor/src/go/number_expr.go
package main

// NumberExpr is a Concrete Element — a literal number.
type NumberExpr struct {
	value float64
}

func NewNumberExpr(value float64) *NumberExpr {
	return &NumberExpr{value: value}
}

func (n *NumberExpr) Value() float64 { return n.value }

func (n *NumberExpr) Accept(visitor Visitor) {
	visitor.VisitNumber(n) // double dispatch
}

File: 26-Visitor/src/rust/number_expr.rs
use crate::expression::Expression;
use crate::visitor::Visitor;

/// Concrete Element — a literal number.
pub struct NumberExpr {
    value: f64,
}

impl NumberExpr {
    pub fn new(value: f64) -> Self {
        NumberExpr { value }
    }

    pub fn value(&self) -> f64 {
        self.value
    }
}

impl Expression for NumberExpr {
    fn accept(&self, visitor: &mut dyn Visitor) {
        visitor.visit_number(self); // double dispatch
    }
}
